tiledb/sm/filesystem/s3.cc

```diff
@@ -947,40 +976,53 @@ Status S3::init_client() const {
   }
 #endif
 
+  if (!aws_access_key_id.empty() && !aws_secret_access_key.empty() && !aws_role_arn.empty()){
+    return Status::S3Error("Authentication credentials can exclusively be either long lived access credentials"
+        "or temporal ones assumed by the role");
+  }
+
   // If the user set config variables for AWS keys use them.
   if (!aws_access_key_id.empty() && !aws_secret_access_key.empty()) {
     Aws::String access_key_id(aws_access_key_id.c_str());
     Aws::String secret_access_key(aws_secret_access_key.c_str());
-    client_creds_ = std::unique_ptr<Aws::Auth::AWSCredentials>(
-        new Aws::Auth::AWSCredentials(access_key_id, secret_access_key));
-
-    // If the user has set a session token (for AWS Security Token Service)
-    // then use it:
-    //     - https://docs.aws.amazon.com/STS/latest/APIReference/Welcome.html
-    // For testing run: `aws sts get-session-token --duration-seconds 900`. See:
-    //     -
-    //     https://docs.aws.amazon.com/cli/latest/reference/sts/get-session-token.html
-    if (!aws_session_token.empty()) {
-      Aws::String session_token(aws_session_token.c_str());
-      client_creds_->SetSessionToken(session_token);
-    }
+    Aws::String session_token(
+        !aws_session_token.empty() ? aws_session_token.c_str() : "");
+    credentials_provider_ =
+        std::make_shared<Aws::Auth::SimpleAWSCredentialsProvider>(
+            access_key_id, secret_access_key, session_token);
   }
 
-  if (client_creds_ == nullptr) {
-    client_ = Aws::MakeShared<Aws::S3::S3Client>(
-        constants::s3_allocation_tag.c_str(),
-        *client_config_,
-        Aws::Client::AWSAuthV4Signer::PayloadSigningPolicy::Never,
-        use_virtual_addressing_);
-  } else {
+  // If AWS Role ARN provided instead of access_key and secret_key,
+  // temporary credentials will be fetched by assuming this role.
+  else if (!aws_role_arn.empty()) {
```

Comment:
We typically keep the `else if {` statement on the same line as the closing brace. Let's do that and move the block comment above into this if-statement:
```
} else if {
  // The AWS Role ARN was provided instead ...
```